great_expectations/experimental/datasources/dynamic_pandas.py

```diff
@@ -0,0 +1,284 @@
+    "StorageOptions",
+}
+
+NEED_SPECIAL_HANDLING: Dict[str, Set[str]] = defaultdict(set)
+FIELD_SKIPPED_UNSUPPORTED_TYPE: Set[str] = set()
+FIELD_SKIPPED_NO_ANNOTATION: Set[str] = set()
+
+
+class _SignatureTuple(NamedTuple):
+    name: str
+    signature: inspect.Signature
+
+
+class _FieldSpec(NamedTuple):
+    type: Type
+    default_value: object  # ... for required value
+
+
+FIELD_SUBSTITUTIONS: Final[Dict[str, Dict[str, _FieldSpec]]] = {
+    # CSVAsset
+    "filepath_or_buffer": {"path": _FieldSpec(pathlib.Path, ...)},
+    # JSONAsset
+    "path_or_buf": {"path": _FieldSpec(pathlib.Path, ...)},
+    # misc
+    "filepath": {"path": _FieldSpec(pathlib.Path, ...)},
+    "dtype": {"dtype": _FieldSpec(Optional[dict], None)},  # type: ignore[arg-type]
+    "dialect": {"dialect": _FieldSpec(Optional[str], None)},  # type: ignore[arg-type]
+    "usecols": {"usecols": _FieldSpec(Union[int, str, Sequence[int], None], None)},  # type: ignore[arg-type]
+    "skiprows": {"skiprows": _FieldSpec(Union[Sequence[int], int, None], None)},  # type: ignore[arg-type]
+}
+
+_METHOD_TO_CLASS_NAME_MAPPINGS: Final[Dict[str, str]] = {
+    "csv": "CSVAsset",
+    "json": "JSONAsset",
+    "html": "HTMLAsset",
+}
+
+_TYPE_REF_LOCALS: Final[Dict[str, Type]] = {
+    "Literal": Literal,  # type: ignore[dict-item]
+    "Sequence": Sequence,
+    "Hashable": Hashable,
+    "Iterable": Iterable,
+    "FilePath": FilePath,
+    "Pattern": re.Pattern,
+    "CSVEngine": CSVEngine,
+    "IndexLabel": IndexLabel,
+    "CompressionOptions": CompressionOptions,
+    "StorageOptions": StorageOptions,
+}
+
+# TODO: make these functions a generator pipeline
+
+
+def _extract_io_methods(
+    whitelist: Optional[Sequence[str]] = None,
+) -> List[Tuple[str, DataFrameFactoryFn]]:
+    member_functions = inspect.getmembers(pd, predicate=inspect.isfunction)
+    if whitelist:
+        return [t for t in member_functions if t[0] in whitelist]
+    return [t for t in member_functions if t[0].startswith("read_")]
+
+
+def _extract_io_signatures(
+    io_methods: List[Tuple[str, DataFrameFactoryFn]]
+) -> List[_SignatureTuple]:
+    signatures = []
+    for name, method in io_methods:
+        sig = inspect.signature(method)
+        signatures.append(_SignatureTuple(name, sig))
+    return signatures
+
+
+def _get_default_value(
+    param: inspect.Parameter,
+) -> object:
+    if param.default is inspect.Parameter.empty:
+        default = ...
+    else:
+        default = param.default
+    return default
+
+
+def _get_annotation_type(param: inspect.Parameter) -> Union[Type, str, object]:
+    """
+    https://docs.python.org/3/howto/annotations.html#manually-un-stringizing-stringized-annotations
+    """
+    # TODO: parse the annotation string
+    annotation = param.annotation
+    if not isinstance(annotation, str):
+        logger.debug(type(annotation), annotation)
+        return annotation
+
+    types: list = []
+
+    union_parts = annotation.split(" | ")
```

Comment:
I agree that's safer, I was trying to avoid extra operations.
My thought was that since pandas is using postponed annotations and `X | Y` the whitespace around the `|` is always going to be consistent.

What I was worried about is that if they forget to use their convention somewhere and we get a `Union[X, Y]` then this will fail.